Enable FEATURE_MULTIREG_RET for x86 RyuJIT

adiaaida opened the pull request:
This change enables the following:

1) Enable FEATURE_MULTIREG_RET for x86 RyuJIT, which is used for calls
with long return types.

2) Enable ReturnTypeDesc for x86 RyuJIT to describe the return types for
longs. Included in this is setting up the correct number of return
registers for longs on x86, and setting up the correct registers for long
returns. This is required for enabling impFixupCallLongReturn.

3) Enabling HasMultiRegRetVal for longs on x86.

4) Enabling impFixupCallLongReturn and impAssignStructClassToVar for x86
longs. Sets up the call's ReturnTypeDesc and sets lvIsMultiRegArgOrRet for
long's tmp.

adiaaida commented:
@sivarv @BruceForstall @CarolEidt PTAL
/cc: @dotnet/jit-contrib

CarolEidt commented on src/jit/compiler.cpp:
```diff
@@ -540,7 +540,7 @@ var_types    Compiler::argOrReturnTypeForStruct(unsigned size, CORINFO_CLASS_HAN
 
     default:
-#if FEATURE_MULTIREG_RET
+#if FEATURE_MULTIREG_RET && !defined(_TARGET_X86_)
```
This is a bit awkward.  I think that this method should be refactored, as it is now quite difficult to follow the logic. Perhaps the 32-bit architectures could at least be separated from the 64-bit ones.

RussKeldorph replied:
If you don't change this, you need to update the corresponding `#endif` comment.

adiaaida replied:
Actually, most of this code doesn't make sense to me in the first place since FEATURE_MULTIREG_RET isn't set for _TARGET_ARM64_, so except for the ifs, none of this ever gets executed. The reason I had to add the !_TARGET_X86_ is because MAX_RET_MULTIREG_BYTES isn't set for x86. I could update that (to be 0?) and then this code wouldn't change.

CarolEidt replied:
MAX_RET_MULTIREG_BYTES should be set to 2 for x86, since it uses 2 registers to return longs (right?) The problem is that this method is assuming that a multireg return is a struct. That should either be fixed, or the code should be factored into a 32-bit version and a 64-bit version (which I think would be cleaner).

briansull replied:
I am current working on implementing the FEATURE_MULTIREG_RET for ARM64

When FEATURE_MULTIREG_RET is defined for a target the value of MAX_RET_MULTIREG_BYTES should also be updated.  This value is just the maximum size that we would ever see  for any struct that is returned in multiple registers, so for x86 that would probably be 8 (or maybe 0).  Setting this to 8 doesn't mean that all 8 byte structs or other 8 byte types have to be returned in registers.  It would just mean that there is some 8 byte type or struct that is returned in registers.  I believe that TYP_LONG on x86 is return in two registers, but 8 byte structs are not returned in two registers.

As the code already does the proper thing here, simply adding a comment about x86 behavior is sufficient:

// For _TARGET_X86_ we never pass or return a TYP_STRUCT type using multiple registers.

CarolEidt commented on src/jit/gentree.h:
```diff
@@ -2399,12 +2399,23 @@ struct ReturnTypeDesc
+        m_inited = true;
+#endif
+
+#else
```
Since the plan is to universally use a ReturnTypeDesc on calls (even single-reg), I think you should use the same constructor behavior, and then initialize it.

adiaaida replied:
@sivarv suggested I change it this way, and I didn't realize that we were planning on using ReturnTypeDesc for all calls. I can change it back to using Initialize, but that causes some issues when we morph treenodes into helper calls (though Siva and I discussed that we may need a Reinitialize function for those cases).

CarolEidt replied:
Look at the TODO-AllArch comment on line 2528.
In general, I don't like having off-by-one behavior when it's not necessary, but most especially when it is temporary.
I would not recommend having a Reinitialize function when a node is morphed - it should instead have a "change" or "morph" verb in its name.

sivarv replied:
How about adding a function to reset the state of ReturnTypeDesc to un-initialized state?

e.g. 

```
void Reset()
{
   m_regType0 = TYP_UNKNOWN;
   m_regType1 = TYP_UNKNOWN;

#ifdef DEBUG
     m_inited = false;
#endif
}
```

Now wherever we need to re-initialize (e.g. fgMorphIntoHelperCall())
retTypeDesc->Reset();
retTypeDesc->Initialize()

Note that the default constructor of ReturnTypeDesc can also use Reset() to 

e.g.

```
ReturnTypeDesc()
{
   Reset();
}
```

sivarv replied:
@adiaaida - even if we initialize ReturnTypeDesc in default constructor, still it will not solve the issue of fgMorphIntoHelperCall(). In that particular case, an existing gentree node was getting morph'ed into a GT_CALL node by changing its oper and other fields.

CarolEidt commented on src/jit/gentree.h:
```diff
@@ -2516,7 +2526,7 @@ struct GenTreeCall final : public GenTree
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+#ifdef FEATURE_MULTIREG_RET
     ReturnTypeDesc    gtReturnTypeDesc;
 
```
Changing this to FEATURE_MULTIREG_RET will enable this for arm64 as well, right? Is that the right behavior at this point?

adiaaida replied:
Arm64 currently doesn't have FEATURE_MULTIREG_RET set, so these changes will not currently affect Arm64. I know Brian is working on enabling FEATURE_MULTIREG_RET for Arm64, so once he does, these functions will become available to Arm64.

CarolEidt commented on src/jit/importer.cpp:
```diff
@@ -7111,6 +7116,60 @@ GenTreePtr                Compiler::impFixupCallStructReturn(GenTreePtr     call
+//  impFixupCallLongReturn: For a call node that returns a struct type either
+//  adjust the return type to an enregisterable type, or set the flag to indicate
+//  struct return via retbuf arg.
+//
```
This comment needs to be updated.

adiaaida replied:
Thanks for the note. Will update.

briansull commented on src/jit/target.h:
```diff
@@ -387,7 +387,13 @@ typedef unsigned short          regPairNoSmall; // arm: need 12 bits
   #define FEATURE_MULTIREG_RET          0  // Support for returning a single value in more than one register  
+#else
+  #define FEATURE_MULTIREG_RET          1  // Support for returning a single value in more than one register  
+#endif
```
If either FEATURE_MULTIREG_RET  or FEATURE_MULTIREG_ARGS  is enabled (by setting them to 1) then the value of FEATURE_MULTIREG_ARGS_OR_RET  must also be set to 1.

Probably best to expand this #ifdef on LEGACY backend to cover all three of these value:

adiaaida replied:
done.

sivarv commented on src/jit/importer.cpp:
```diff
@@ -7111,6 +7116,58 @@ GenTreePtr                Compiler::impFixupCallStructReturn(GenTreePtr     call
+//    Returns new GenTree node after fixing long return of call node
+//
+GenTreePtr                Compiler::impFixupCallLongReturn(GenTreePtr     call,
+                                                             CORINFO_CLASS_HANDLE retClsHnd)
```
(Minor) alignment of second param with the first.

sivarv commented on src/jit/importer.cpp:
```diff
@@ -6827,11 +6827,16 @@ var_types           Compiler::impImportCall (OPCODE         opcode,
+        {
+            call = impFixupCallLongReturn(call, sig->retTypeClass);
+        }
+
```
We could avoid redundant condition checking by writing it as follows:

```
if (call->gtOper == GT_CALL)
{
    if (varTypeIsStruct(callRetType))
    {
          call = impFixupCallStructReturn(call, sig->retTypeClass);
    }
    else if (varTypeIsLong(callRetTyp))
    {
          call = impFixupCallLongReturn(call, sig->retTypeClass);
    }
}
```

sivarv commented on src/jit/importer.cpp:
```diff
@@ -7111,6 +7116,58 @@ GenTreePtr                Compiler::impFixupCallStructReturn(GenTreePtr     call
+        // No need to assign a multi-reg long to a local var if:
+        //  - It is a tail call or 
+        //  - The call is marked for in-lining later
+        return impAssignStructClassToVar(call, retClsHnd);
```
How come long ret fix-up path is using a method meant for structs?  Would it make sense to rename the routine as impAssignMultiRegTypeToVar()?  

Based on source search of all calls to impAssignStructClassToVar(), i see this is currently getting used by Arm and x64 unix for multi-reg returned structs.  Now we intend to use it for multi-reg returned long types.

CarolEidt replied:
I agree with the idea of renaming this as @sivarv suggested.

sivarv commented on src/jit/compiler.h:
```diff
@@ -2644,6 +2644,9 @@ protected :
                                                  CORINFO_CLASS_HANDLE retClsHnd);
 
+    GenTreePtr          impFixupCallLongReturn(GenTreePtr           call,
+                                                 CORINFO_CLASS_HANDLE retClsHnd);
```
(Minor) alignment of second param with the first.

sivarv commented on src/jit/gentree.cpp:
```diff
@@ -13832,6 +13832,16 @@ regNumber ReturnTypeDesc::GetABIReturnReg(unsigned idx)
         }
     }
+
+#elif defined(_TARGET_X86_) && !defined(LEGACY_BACKEND)
```
Here we don't need !defined(LEGACY_BACKEND), since this routine is correct even for legacy backend.

adiaaida commented:
I updated and I think I addressed all the comments up to this point. I decided that the easiest thing with the change I'd made in compiler.cpp would be to set MAX_RET_MULTIREG_BYTES to 0, as, like Brian said, this is used for structs being returned in multiple registers, which we don't do for x86 right now.

CarolEidt commented:
I strongly disagree with setting MAX_RET_MULTIREG_BYTES to 0. The whole point of renaming FEATURE_MULTIREG_STRUCTS to FEATURE_MULTIREG_ARGS_OR_RET was to work toward unifying the handling of multi-register args and returns. This moves us in the wrong direction.

sivarv commented on src/jit/gentree.cpp:
```diff
@@ -13758,11 +13759,14 @@ void ReturnTypeDesc::Initialize(Compiler* comp, CORINFO_CLASS_HANDLE retClsHnd)
         }
     }
 
+#elif defined(_TARGET_X86) && !defined(LEGACY_BACKEND)
```
Here we don't need second part of the condition. The way to look at is that ReturnTypeDesc implementation is for x86 and right now it is used by RyuJIT.

briansull commented:
I believe that it should be the case that it already works if MAX_RET_MULTIREG_BYTES is set to 8.
The only reference that we have to this define is here:

#if FEATURE_MULTIREG_RET
        if (forReturn)
        {
            if (size <= MAX_RET_MULTIREG_BYTES)
            {
#ifdef _TARGET_ARM64_
                // TODO-ARM64-HFA - Implement x0,x1 returns   
                // TODO-ARM64     - Implement HFA returns   
#endif // _TARGET_XXX_
            }
        }
#endif // FEATURE_MULTIREG_RET

sivarv commented on src/jit/gentree.cpp:
```diff
@@ -13758,11 +13759,14 @@ void ReturnTypeDesc::Initialize(Compiler* comp, CORINFO_CLASS_HANDLE retClsHnd)
 
+#elif defined(_TARGET_X86_)
+    m_regType0 = TYP_INT;
+    m_regType1 = TYP_INT;
```
Here we are making an assumption on x86 that ReturnTypeDesc is always used for TYP_LONG return types.  But eventually we want to use ReturnTypeDesc for describing all return types and not just long.

How easy or difficult to check that the type handle passed corresponds to that of TYP_LONG/TYP_ULONG?
I would suggest minimum to add a Assumption/TODO comment here.

sivarv commented on src/jit/importer.cpp:
```diff
@@ -7111,6 +7118,59 @@ GenTreePtr                Compiler::impFixupCallStructReturn(GenTreePtr     call
+GenTreePtr                Compiler::impFixupCallLongReturn(GenTreePtr     call,
+                                                           CORINFO_CLASS_HANDLE retClsHnd)
+{
+#if defined(_TARGET_X86_) && !defined(LEGACY_BACKEND)
```
It would be a good idea to add a comment line here why this logic is confined to RyuJIt alone.

sivarv commented on src/jit/target.h:
```diff
@@ -385,9 +385,17 @@ typedef unsigned short          regPairNoSmall; // arm: need 12 bits
+  #define FEATURE_MULTIREG_ARGS_OR_RET  1  // Support for passing and/or returning single values in more than one register
+  #define FEATURE_MULTIREG_ARGS         0  // Support for passing a single argument in more than one register  
+  #define FEATURE_MULTIREG_RET          1  // Support for returning a single value in more than one register  
+  #define MAX_RET_MULTIREG_BYTES        0  // Maximum size of a struct that could be returned in more than one register
```
I agree with Carol here that this should be defined as 8 instead of zero.

sivarv commented on src/jit/gentree.h:
```diff
@@ -4009,7 +4016,7 @@ struct GenTreeCopyOrReload : public GenTreeUnOp
         }
 
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+#if FEATURE_MULTIREG_RET
```
In gentree.cpp there is another function GenTreeCall::GetOtherRegMask() that is implemented only for FEATURE_UNIX_AMD64_STRUCT_PASSING and that should be changed to FEATURE_MULTIREG_RET

sivarv commented on src/jit/gentree.h:
```diff
@@ -2826,6 +2831,8 @@ struct GenTreeCall final : public GenTree
     { 
 #ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
         return varTypeIsStruct(gtType) && !HasRetBufArg(); 
+#elif defined(_TARGET_X86_) && !defined(LEGACY_BACKEND)
```
Here too condition "!defined(LEGACY_BACKEND)" is not needed.

sivarv replied:
TODO comment line on 2829 can be deleted now?

sivarv replied:
On second thought, i think it would make sense to have second condition "!defined(LEGACY_BACKEND)" since it is used in flowgraph, importer, morph and lsra.cpp.

adiaaida replied:
There are multiple places where HasMultiRegRetVal is called that, I think, could potentially include LEGACY_BACKEND. Are we sure we want LEGACY_BACKEND to potentially return true in this function?

sivarv commented on src/jit/gentree.h:
```diff
@@ -3964,7 +3971,7 @@ struct GenTreeCopyOrReload : public GenTreeUnOp
     // TODO-ARM: Enable this when multi-reg call node support is added.
     // TODO-X86: Enable this when multi-reg call node support is added.
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+#if FEATURE_MULTIREG_RET
```
We can delete both the TODO comments on lines 3972 and 3973

sivarv commented on src/jit/gentree.h:
```diff
@@ -3981,7 +3988,7 @@ struct GenTreeCopyOrReload : public GenTreeUnOp
     void ClearOtherRegs()
     {
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+#if FEATURE_MULTIREG_RET
```
We can delete both the TODO comment lines on 3987 and 3988

sivarv commented on src/jit/gentree.h:
```diff
@@ -2602,7 +2607,7 @@ struct GenTreeCall final : public GenTree
             gtRegNum = reg;
         }
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+#if FEATURE_MULTIREG_RET
```
Both the TODO comments can be deleted now (lines 2600 and 2601)

sivarv commented on src/jit/gentree.h:
```diff
@@ -2575,7 +2580,7 @@ struct GenTreeCall final : public GenTree
         }
 
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+#if FEATURE_MULTIREG_RET
```
Both the TODO comment lines (2572 and 2573) can be deleted now.

sivarv commented on src/jit/gentree.h:
```diff
@@ -2547,7 +2552,7 @@ struct GenTreeCall final : public GenTree
     ReturnTypeDesc*   GetReturnTypeDesc()
     {
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+#if FEATURE_MULTIREG_RET
```
TODO comments lines 2550 and 2551 can be deleted now.
Also the NOTE: can be deleted now.

sivarv commented on src/jit/gentree.h:
```diff
@@ -2516,7 +2521,7 @@ struct GenTreeCall final : public GenTree
     // TODO-X86: enable this for long returns on x86
     // TODO-AllArch: enable for all call nodes to unify single-reg and multi-reg returns.
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+#if FEATURE_MULTIREG_RET
```
TODO lines 2521 and 2522 can be deleted now.

adiaaida commented:
I think I have fixed the last of the comments up to this point.

sivarv commented on src/jit/gentree.h:
```diff
@@ -3962,9 +3960,7 @@ struct GenTreeCopyOrReload : public GenTreeUnOp
     // First register is is always given by gtRegNum.
     // Currently enabled for x64 unix.
```
(Minor) the above comment line can be deleted now.

sivarv commented:
Looks good.

CarolEidt commented:
Looks good - thanks!
